### PR TITLE
Make Sure Newly Spawned Cells Survive the Night

### LINKED ISSUE #3893: Cells spawning at night / close to night should be given extra resources
to ensure that the cells can survive so that there isn't mass dying around the player during the night

### PR DESCRIPTION
**Pull request by GameDungeon**
**Brief Description of What This PR Does**
Currently, this adds the exact additional glucose a cell needs to survive the night, plus the default 30 seconds of glucose it gets. 

I wasn't sure the best way to go about this, so it would probably be best if HH reviewed the code before this gets merged. Additionally, this is a balancing issue (Currently I give the exact glucose needed to survive the night to any cell spawns during or right before the night. This is probably not how it should work.)

I'm also a little afraid this might break something in a later stage if some testers could look at that. 

**Related Issues**

Closes #3893

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by buckly90** [comment]
I like the idea, but feel it should be less of a complete failsafe and more of a buffer to create more room for error. That is, that players that are due to starve before reproducing shouldn't get awarded enough glucose to survive the night on it's own, but enough that if they were just short, that they'll then be able to survive.

Forgiving, but not rewarding.

When I get home tonight, I'll see about playing around with this to see if there's a good amount to be struck.

**Comment by GameDungeon** [comment]
The dev center is doing some weird things and I'm not sure why, but I'm pretty sure the failed build isn't the PR's fault.

**Comment by buckly90** [comment]
Seems to work great. A subtle boost to available glucose that allows players to survive the night when stationary. Not substantial enough to necessarily be noticeable beyond that.

It may be worth changing the INITIAL_COMPOUND_NIGHT_MULTIPLIER to 1.1 for an additional amount of buffer in the event players move a bit for some reason as a slow plant cell.

**Comment by hhyyrylainen** [comment]
What's the status here? It seems like the code checks pass and @buckly90 has done playtesting on this. Is this just waiting for a code review, then?

File: src/microbe_stage/Microbe.Interior.cs
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.Interior.cs:387]: I think this comment is insufficient to explain the complexity that is the next two lines, which I also think should be put in as a method in DayNightCycle to concentrate the tricky calculations there.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/microbe_stage/Microbe.Interior.cs
Comment (hhyyrylainen) [inline src/microbe_stage/Microbe.Interior.cs:399]: Also just noticed as I opened this in my IDE that there's a typo here (`remaning`).

File: simulation_parameters/Constants.cs
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:35]: Typo in `inital`

**Comment by GameDungeon** [comment]
It was mostly waiting on a code review, plus validation from you due to the weirdness here. I'll go check the review when I get home :) Thanks

File: src/microbe_stage/Microbe.Interior.cs
Comment (GameDungeon) [reply src/microbe_stage/Microbe.Interior.cs:396]: I don't think a cache works, as this is a dynamic value based on the time of day.

Comment (hhyyrylainen) [reply src/microbe_stage/Microbe.Interior.cs:396]: You mean the calculated compound balance changes? Still from small lag spike viewpoint it's not a good design. I commented later that some kind of more in-depth check could be made to calculate which compounds vary during the night. I see that as potentially much more useful calculation result that can be cached and from it it could be just a quick calculation to determine the resources to give to each spawned cell.

File: src/microbe_stage/Microbe.Interior.cs
Comment (GameDungeon) [reply src/microbe_stage/Microbe.Interior.cs:391]: DaytimeFraction is called from world generation settings, so the nightTimeFraction call would have to be added there, which is a weird place for it. Your call on that one.

Comment (hhyyrylainen) [reply src/microbe_stage/Microbe.Interior.cs:391]: I meant like that you'd have `settings.NightTimeFraction` property that is literally just `=> 1 - DaytimeFraction`. That way no one can accidentally typo and incorrectly calculate night time fraction someplace in the code.

Comment (GameDungeon) [reply src/microbe_stage/Microbe.Interior.cs:391]: I guess it could be added to day night cycle actually, it just would need to be called from a different places then DaytimeFraction normally is.

Comment (hhyyrylainen) [reply src/microbe_stage/Microbe.Interior.cs:391]: Well I can't remember at this point the context of what I reviewed 3 weeks ago, so just see what makes sense to you. I'll re-review this later then.

File: simulation_parameters/Constants.cs
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:37]:
```suggestion
    ///   This is a fraction of the day.
```
This is not a percentage. I want to keep percentages and fractions really clear in the code which is which as otherwise there's going to be massive bugs.

File: simulation_parameters/Constants.cs
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:44]: Same here.

File: src/general/DayNightConfiguration.cs
Comment (hhyyrylainen) [inline src/general/DayNightConfiguration.cs:28]:
```suggestion
    ///   Fraction of the in-game day which does not have sunlight.
```
I think that's a typo.

**Comment by GameDungeon** [comment]
Given this was pre-ECS and there now exists much better ways to approach this, I'm going to close this.

**Pull request closed without merging**
